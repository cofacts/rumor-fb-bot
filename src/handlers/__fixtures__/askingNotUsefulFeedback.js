export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
  },
};
