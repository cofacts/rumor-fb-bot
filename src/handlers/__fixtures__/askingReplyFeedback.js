export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};

export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const articleData = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};
