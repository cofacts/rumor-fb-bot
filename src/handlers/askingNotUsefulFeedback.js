import gql from '../gql';
import { getArticleURL, createPostbackAction } from './utils';

/**
 * The state that a user is telling us why she thinks our replies are not useful.
 * We also ask for confirmation.
 */
export default async function askingNotUsefulFeedback(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;

  if (!data.selectedReplyId) {
    throw new Error('selectedReply not set in data');
  }

  if (event.input === 'n') {
    const {
      data: {
        action: { feedbackCount },
      },
    } = await gql`
      mutation(
        $comment: String!
        $vote: FeedbackVote!
        $articleId: String!
        $replyId: String!
      ) {
        action: CreateOrUpdateArticleReplyFeedback(
          comment: $comment
          articleId: $articleId
          replyId: $replyId
          vote: $vote
        ) {
          feedbackCount
        }
      }
    `(
      {
        articleId: data.selectedArticleId,
        replyId: data.selectedReplyId,
        comment: 'none',
        vote: 'DOWNVOTE',
      },
      { userId }
    );

    replies = [
      {
        type: 'text',
        content: {
          text:
            // ? We've received feedback from you and other {count - 1} person(s)!
            // : Thanks. You're the first one who gave feedback on this reply!
            feedbackCount > 1
              ? `感謝您與其他 ${feedbackCount - 1} 人的回饋。`
              : '感謝您的回饋，您是第一個評論這個回應的人 :)',
        },
      },
      {
        type: 'text',
        content: {
          // If you have something to say about this article,
          // feel free to submit us your own reply!
          text: `💁 若您認為自己能回應得更好，歡迎到 ${getArticleURL(
            data.selectedArticleId
          )} 提交新的回應唷！`,
        },
      },
    ];
    state = '__INIT__';
  } else {
    data.comment = event.input;

    replies = [
      {
        type: 'text',
        content: {
          // The following is your reason: {reason}
          text: `以下是您所填寫的理由：「${event.input}」`,
        },
      },
      {
        type: 'buttons',
        content: {
          attachment: {
            type: 'template',
            payload: {
              template_type: 'button',
              // Fact checkers will see why you find this reply not helpful.
              // Please confirm.
              text: '我們會把您覺得回應沒幫助的原因呈現給編輯們看。請確認：',
              buttons: [
                // OK. Submit now!
                createPostbackAction('明白，我要送出', 'y'),
                // Revise my reason
                createPostbackAction('重寫送出的理由', 'r'),
                // Skip
                createPostbackAction('算了，我不想填', 'n'),
              ],
            },
          },
        },
      },
    ];

    state = 'ASKING_NOT_USEFUL_FEEDBACK_SUBMISSION';
  }
  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
