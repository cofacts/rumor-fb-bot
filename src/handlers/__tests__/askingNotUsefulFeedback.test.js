jest.mock('../../gql');

import askingNotUsefulFeedback from '../askingNotUsefulFeedback';
import * as apiResult from '../__fixtures__/askingNotUsefulFeedback';
import gql from '../../gql';

const commonParamsNone = {
  data: {
    searchedText: 'è²¼åœ–',
    foundArticleIds: [
      'AWDZYXxAyCdS-nWhumlz',
      '5483323992880-rumor',
      'AV-Urc0jyCdS-nWhuity',
      'AVsh8u7StKp96s659Dgq',
    ],
    selectedArticleId: 'AWDZYXxAyCdS-nWhumlz',
    foundReplyIds: ['AWDZeeV0yCdS-nWhuml8'],
    selectedReplyId: 'AWDZeeV0yCdS-nWhuml8',
  },
  state: 'ASKING_NOT_USEFUL_FEEDBACK',
  event: {
    type: 'postback',
    input: 'n',
    timestamp: 1519019734813,
    postback: { data: '{"input":"n","issuedAt":1519019701265}' },
  },
  issuedAt: 1519019735467,
  userId: 'Uaddc74df8a3a176b901d9d648b0fc4fe',
  replies: undefined,
  isSkipUser: false,
};

it('handles "none" postback with no other existing feedbacks', async () => {
  gql.__push(apiResult.oneFeedback);

  expect(await askingNotUsefulFeedback(commonParamsNone)).toMatchSnapshot();
});

it('handles "none" postback with other existing feedbacks', async () => {
  gql.__push(apiResult.twoFeedbacks);

  expect(await askingNotUsefulFeedback(commonParamsNone)).toMatchSnapshot();
});

const commonParamsComment = JSON.parse(JSON.stringify(commonParamsNone));
commonParamsComment.event = {
  type: 'text',
  input: 'comment',
  timestamp: 1519019734813,
  postback: { data: '{"input":"comment","issuedAt":1519019701265}' },
};

it('handles text comment with other existing feedbacks', async () => {
  expect(await askingNotUsefulFeedback(commonParamsComment)).toMatchSnapshot();
});

const commonParamsNoId = JSON.parse(JSON.stringify(commonParamsNone));
commonParamsNoId.data.selectedReplyId = undefined;

it('handles undefined reply id', () => {
  expect(askingNotUsefulFeedback(commonParamsNoId)).rejects.toThrow(
    'selectedReply not set in data'
  );
});

afterEach(() => {
  gql.__reset();
});
