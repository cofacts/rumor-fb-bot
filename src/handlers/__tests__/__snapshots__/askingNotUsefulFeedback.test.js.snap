// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles "none" postback with no other existing feedbacks 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AWDZYXxAyCdS-nWhumlz",
      "5483323992880-rumor",
      "AV-Urc0jyCdS-nWhuity",
      "AVsh8u7StKp96s659Dgq",
    ],
    "foundReplyIds": Array [
      "AWDZeeV0yCdS-nWhuml8",
    ],
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": "n",
    "postback": Object {
      "data": "{\\"input\\":\\"n\\",\\"issuedAt\\":1519019701265}",
    },
    "timestamp": 1519019734813,
    "type": "postback",
  },
  "isSkipUser": false,
  "issuedAt": 1519019735467,
  "replies": Array [
    Object {
      "content": Object {
        "text": "感謝您的回饋，您是第一個評論這個回應的人 :)",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "💁 若您認為自己能回應得更好，歡迎到 https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz 提交新的回應唷！",
      },
      "type": "text",
    },
  ],
  "state": "__INIT__",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;

exports[`handles "none" postback with other existing feedbacks 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AWDZYXxAyCdS-nWhumlz",
      "5483323992880-rumor",
      "AV-Urc0jyCdS-nWhuity",
      "AVsh8u7StKp96s659Dgq",
    ],
    "foundReplyIds": Array [
      "AWDZeeV0yCdS-nWhuml8",
    ],
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": "n",
    "postback": Object {
      "data": "{\\"input\\":\\"n\\",\\"issuedAt\\":1519019701265}",
    },
    "timestamp": 1519019734813,
    "type": "postback",
  },
  "isSkipUser": false,
  "issuedAt": 1519019735467,
  "replies": Array [
    Object {
      "content": Object {
        "text": "感謝您與其他 1 人的回饋。",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "💁 若您認為自己能回應得更好，歡迎到 https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz 提交新的回應唷！",
      },
      "type": "text",
    },
  ],
  "state": "__INIT__",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;

exports[`handles text comment with other existing feedbacks 1`] = `
Object {
  "data": Object {
    "comment": "comment",
    "foundArticleIds": Array [
      "AWDZYXxAyCdS-nWhumlz",
      "5483323992880-rumor",
      "AV-Urc0jyCdS-nWhuity",
      "AVsh8u7StKp96s659Dgq",
    ],
    "foundReplyIds": Array [
      "AWDZeeV0yCdS-nWhuml8",
    ],
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": "comment",
    "postback": Object {
      "data": "{\\"input\\":\\"comment\\",\\"issuedAt\\":1519019701265}",
    },
    "timestamp": 1519019734813,
    "type": "text",
  },
  "isSkipUser": false,
  "issuedAt": 1519019735467,
  "replies": Array [
    Object {
      "content": Object {
        "text": "以下是您所填寫的理由：「comment」",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "buttons": Array [
              Object {
                "payload": "y",
                "title": "明白，我要送出",
                "type": "postback",
              },
              Object {
                "payload": "r",
                "title": "重寫送出的理由",
                "type": "postback",
              },
              Object {
                "payload": "n",
                "title": "算了，我不想填",
                "type": "postback",
              },
            ],
            "template_type": "button",
            "text": "我們會把您覺得回應沒幫助的原因呈現給編輯們看。請確認：",
          },
          "type": "template",
        },
      },
      "type": "buttons",
    },
  ],
  "state": "ASKING_NOT_USEFUL_FEEDBACK_SUBMISSION",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;
